internal/controller/controllers/clusterdeployments_controller.go

```diff
@@ -138,6 +138,9 @@ func (r *ClusterDeploymentsReconciler) Reconcile(ctx context.Context, req ctrl.R
 		if err != nil {
 			return r.updateStatus(ctx, clusterDeployment, cluster, err)
 		}
+		if err = r.NotifyClusterDeploymentAgents(req.NamespacedName); err != nil {
```

Comment:
I'm not understanding how this solves the problem of identifying when a ClusterDeployment has been deleted. Can you describe the pattern?

Reply:
Allow me to explain:
The basic idea I had in mind is that we should trigger an agent reconcile in some cases (mentioned in the commit-msg) so the `agent_controller` can determine whether or not the resource should be deleted.
Since those cases start with the `clusterDeployment`, it will naturally first reconcile on that controller, so it needs to "tell" the `agent_controller` that something happened.

The outcome is that I placed (in `clusterdeployments_controller`) a call to notify `agent_controller` about relevant agents in places that indicate the day1 cluster got deregistered (line#135) and on `deregisterClusterIfNeeded()` (line#587), which gets invoked when it gets `k8serrors.IsNotFound`
while trying to get the `clusterDeployment` CR.

On the agent controller side, it starts to reconcile and call `r.deleteAgent` if either `clusterDeployment` or backend cluster is gone; or that agent does not belong to the cluster it is listed for (day2 scenario).

Let me know if that clarifies the current implementation, and how to proceed from here.

Reply:
Thanks for elaborating. The k8s-native way to normally handle this situation is to add a finalizer to the ClusterDeployment. Then at the beginning of the Reconcile, check for a deletion timestamp to be set, and if it is and the finalizer is still there:
* send the reconcile requests to the agent controller
* do any other things you want to do upon deletion
* remove the finalizer and save the resource
* return

I know in our case we have this extra source of truth in the sql DB, but I think it's better to not depend on that state machine as a substitute for the normal kube patterns.

WDYT?

Reply:
Regardless of finilazer  - i don't think that you need to notify the agents when cluster is deleted there is a deletion of the agents in the backend so there should be an event for each agent and it will trigger the reconcile for the agents.

Reply:
Implementation changed to use a finalizer for clusterDeployment deletion.

the cleanup logic `deregisterClusterIfNeeded` is being called to cleanup the backend cluster and delete the agents.

That very same cleanup logic is used when (not changed in this PR):
1. It tries to reconcile for a non-existent `clusterDeployment` - to ensure both the controller and the backend are on the same page.
2. Day1 Cluster gets deleted, which we know won't lead to a `clusterDeployment` removal but to creating a day2 cluster.